### ISSUE TITLE
DicomClient multiple C-Store request cause exception when AsyncOps is not negotiated

### ISSUE THREAD

**Issue by BobSter3000** [fo-dicom/fo-dicom#356]
**Send multiple images using DicomClient send**

**Fails during first image**

**Basic Code Snippet**

```
        string[] files = Directory.GetFiles(@"Images\");
        DicomClient client = new DicomClient();
        client.NegotiateAsyncOps(files.Length, files.Length);

        foreach (string file in files) {

            DicomCStoreRequest request = new DicomCStoreRequest(file);
            request.OnResponseReceived += (req, rsp) => {
                Console.WriteLine(rsp.Status);
            };

            client.AddRequest(request);
        }

        try {
            client.Send("localhost", 4010, false, "DicomSender", "CCDICOM");
        }
        catch (Exception ex) {

            Console.WriteLine(ex.ToString());
        }
```

**Exception from Logging**:

17:34:52  CCDICOM -> Association request:
Calling AE Title:       DicomSender
Called AE Title:        CCDICOM
Remote host:            localhost
Remote port:            4010
Implementation Class:   Implementation Class UID [1.3.6.1.4.1.30071.8]
Implementation Version: fo-dicom 2.9.9
Maximum PDU Length:     16384
Async Ops Invoked:      3
Async Ops Performed:    3
Presentation Contexts:  1
  Presentation Context:  1 [Proposed]
       Abstract Syntax:  Computed Radiography Image Storage
       Transfer Syntax:  JPEG Lossless, Non-Hierarchical, First-Order Prediction (Process 14 [Selection Value 1]): Default Transfer Syntax for Lossless JPEG Image Compression
       Transfer Syntax:  Explicit VR Little Endian
       Transfer Syntax:  Implicit VR Little Endian: Default Transfer Syntax for DICOM
17:34:52  CCDICOM <- Association accept:
Calling AE Title:       DicomSender
Called AE Title:        CCDICOM
Remote host:            localhost
Remote port:            4010
Implementation Class:   Unknown [1.3.6.1.4.1.25403.1.1.1]
Implementation Version: Dicom 0.1
Maximum PDU Length:     116794
Async Ops Invoked:      1
Async Ops Performed:    1
Presentation Contexts:  1
  Presentation Context:  1 [Accept]
       Abstract Syntax:  Computed Radiography Image Storage
       Transfer Syntax:  JPEG Lossless, Non-Hierarchical, First-Order Prediction (Process 14 [Selection Value 1]): Default Transfer Syntax for Lossless JPEG Image Compression
17:34:52  CCDICOM -> C-Store request [1]
17:34:52  Exception processing PDU: System.FormatException: Input string was not in a correct format.
   at System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.Format(String format, Object[] args)
   at Dicom.Network.DicomService.SendNextMessage() in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 900
   at Dicom.Network.DicomService.SendMessage(DicomMessage message) in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 874
   at Dicom.Network.DicomService.SendRequest(DicomRequest request) in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 152
   at Dicom.Network.DicomClient.DicomServiceUser.OnReceiveAssociationAccept(DicomAssociation association) in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomClient.cs:line 507
   at Dicom.Network.DicomService.<ReadAndProcessPDUAsync>d__51.MoveNext() in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 373
17:34:52  Connection closed with error: System.FormatException: Input string was not in a correct format.
   at System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.Format(String format, Object[] args)
   at Dicom.Network.DicomService.SendNextMessage() in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 900
   at Dicom.Network.DicomService.SendMessage(DicomMessage message) in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 874
   at Dicom.Network.DicomService.SendRequest(DicomRequest request) in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 152
   at Dicom.Network.DicomClient.DicomServiceUser.OnReceiveAssociationAccept(DicomAssociation association) in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomClient.cs:line 507
   at Dicom.Network.DicomService.<ReadAndProcessPDUAsync>d__51.MoveNext() in C:\Users\ander\Documents\Visual Studio 2015\Projects\fo-dicom [fo-dicom]\DICOM\Network\DicomService.cs:line 373

**3.0 beta 4 has error** 
**2.0.2 OK**

**Same code 2.0.2**
18:21:46  CCDICOM -> Association request:
Calling AE Title:       DicomSender
Called AE Title:        CCDICOM
Implementation Class:   Implementation Class UID [1.3.6.1.4.1.30071.8]
Implementation Version: fo-dicom 2.0.2
Maximum PDU Length:     16384
Async Ops Invoked:      3
Async Ops Performed:    3
Presentation Contexts:  1
  Presentation Context:  1 [Proposed]
       Abstract Syntax:  Computed Radiography Image Storage
       Transfer Syntax:  JPEG Lossless, Non-Hierarchical, First-Order Prediction (Process 14 [Selection Value 1]): Default Transfer Syntax for Lossless JPEG Image Compression
       Transfer Syntax:  Explicit VR Little Endian
       Transfer Syntax:  Implicit VR Little Endian: Default Transfer Syntax for DICOM
18:21:46  CCDICOM <- Association accept:
Calling AE Title:       DicomSender
Called AE Title:        CCDICOM
Implementation Class:   Unknown [1.3.6.1.4.1.25403.1.1.1]
Implementation Version: Dicom 0.1
Maximum PDU Length:     116794
Async Ops Invoked:      1
Async Ops Performed:    1
Presentation Contexts:  1
  Presentation Context:  1 [Accept]
       Abstract Syntax:  Computed Radiography Image Storage
       Transfer Syntax:  JPEG Lossless, Non-Hierarchical, First-Order Prediction (Process 14 [Selection Value 1]): Default Transfer Syntax for Lossless JPEG Image Compression
18:21:46  CCDICOM -> C-Store request [1]
18:21:46  CCDICOM <- C-Store response [1]: Success
Success
18:21:46  CCDICOM -> C-Store request [2]
18:21:46  CCDICOM <- C-Store response [2]: Success
Success
18:21:46  CCDICOM -> C-Store request [3]
18:21:47  CCDICOM <- C-Store response [3]: Success
Success
18:21:47  CCDICOM -> Association release request
18:21:47  CCDICOM <- Association release response
18:21:47  Connection closed

**Comment by aerik** [comment]
In 2.0.2, calls to SendNextMessage silently return when it's already sending  or when it's waiting for a response to something it's just sent (by way of checking if more stuff is still in _pending than is allowed by the association):

```
 if (_sending) return;
 if (Association.MaxAsyncOpsInvoked > 0 && _pending.Count >= Association.MaxAsyncOpsInvoked) return;
```

Requests are cleared out of _pending in PerformDimse when a response is received

```
if (!DicomMessage.IsRequest(dimse.Type))
{
     var rsp = dimse as DicomResponse;
     lock (_lock)
     {
           var req = _pending.FirstOrDefault(x => x.MessageID == rsp.RequestMessageID);
           if (req != null)
           {
                rsp.UserState = req.UserState;
                req.PostResponse(this, rsp);
                if (rsp.Status.State != DicomState.Pending) _pending.Remove(req);
             }
         }
     return;
 }
```

Basically what happens is that requests are sent serially, each being put off until the response is received for the one before it.  Alternatively, if AsyncOps are negotiated, then that many requests can be sent before the responses are received.

The original behavior works well as long as you receive a response, but sometimes you don't.  My hacked up version of 1.x has a callback for ConnectionClosed to handle cases where no response was received bye DicomClient.

In 3.0 beta 4, instead of returning silently from SendNextMessage, an exception is thrown, basically stopping you from serially sending requests without creating an association for each request.  In the context of async operations, maybe that's the right thing to do, but it breaks serial operations.

**Comment by hashitha** [comment]
@aerik did you manage to get it working on 3.0 beta 4? I am currently sending one image at a time, which is very slow.

**Comment by aerik** [comment]
Well, yeah, I changed it to work like 2.x... (so I build my own DLLs)

**Comment by hashitha** [comment]
When time permits, can you post the file you changed so I can build the DLLs.

**Comment by ztegod** [comment]
In version 1.0 ,if my network is down ,it will throw exception and stop my program.However in version 2.0.2,it will not throw the exception that my connection is stopped,and just stop there without any return!

**Comment by ztegod** [comment]
I have a question to ask you.In version 1.0.1,it will show logs itself.However in version 2.0.2,i need to add a line: 
Dicom.Log.LogManager.SetImplementation(ConsoleLogManager.Instance)
,and then i can see the logs.My confusion is when i dont add this line ,it will not throw exception when my network is down( add multiple CStore requests  and send only once).When i use version 1.0.1,it will throw exception on the screen and stop my program  ,not only shows exceptions in logs (if i dont add the code above,i can see nothing and it just stop there without any reflection).How can i solve it?

**Comment by BobSter3000** [comment]
Returning silently did the trick for me.  I am back to synchronous sending.  

DicomService.cs

                    if (!Options.IgnoreAsyncOps && Association.MaxAsyncOpsInvoked > 0
                        && _pending.Count >= Association.MaxAsyncOpsInvoked)
                    {
                        return; 

If I modified the initial test to IgnoreAsyncOps then the DicomClient would send requests Asynchronously.  Which is not what I intended.

            DicomClient client = new DicomClient();
            //client.NegotiateAsyncOps(files.Length, files.Length);
            DicomServiceOptions options = new DicomServiceOptions();
            options.IgnoreAsyncOps = true;
            client.Options = options;

**Comment by hashitha** [comment]
Thanks @BobSter3000

**Comment by hashitha** [comment]
@BobSter3000 this code doesn't work well with large CTs

```
DicomClient client = new DicomClient();
 //client.NegotiateAsyncOps(files.Length, files.Length);
DicomServiceOptions options = new DicomServiceOptions();
options.IgnoreAsyncOps = true;
client.Options = options;
```

for me, it works fine using the following code

```
DicomClient client = new DicomClient();
client.NegotiateAsyncOps(files.Length, files.Length);
```

tested using Osirix

**Comment by ztegod** [comment]
@BobSter3000 Have you ever removed the code LogManager.SetImplementation(ConsoleLogManager.Instance) in version 2.0.2?I want to know if your exceptions will be thrown to stop your consoleapplication after removing it.By default, logging defaults to the no-op NullLogerManager.However,exceptions should be shown not only in logs,but also to stop our consoleapplictions.I hope you can try it both in version 1.0.1 and 2.0.2.

**Comment by BobSter3000** [comment]
@hashitha I suspect the reason Osirix works is because it is negotiating async ops with fo-dicom.  If asnyc ops is not negotiated then DicomClient cannot send more than one image.  We get the exception.

**Comment by BobSter3000** [comment]
@ztegod I just tried 2.0.2 with a DicomClient echo send.  I put an invalid IP address and did get an exception that could be caught.

**Comment by ztegod** [comment]
@BobSter3000 Im sorry i didnt say clearly.Just like your code above,i have hundreds of images to be sent to c-store scp.However,i try to make the network to be interrupted when im sending images to c-store scp.For example,i have 200 images to be sent,and i make the network to be interrupted in 100th image.In version 2.0.2,if i remove the code:LogManager.SetImplementation(ConsoleLogManager.Instance),my c-store scu will just stop there without any reflection and exception.I hope you can try it again,thanks a lot.

**Comment by hashitha** [comment]
@BobSter3000 I guess that explains it. I didn't know that. Thanks

**Comment by BobSter3000** [comment]
@ztegod I did not try it in 1.0.1, but I do see what you are talking about in 2.0.2 and 3beta4.  What I had always done in a store scu application was to monitor the c-store responses and count the successful against the total.  If there was a difference I would flag the send as failed and refer to the log.  I guess that is why I personally never noticed this condition before.

**Comment by BobSter3000** [comment]
@anders9ustafsson I am trying to find out what the consequences are in reverting back to the 2.0.2 in DicomService.cs.
if (!Options.IgnoreAsyncOps && Association.MaxAsyncOpsInvoked > 0
                    && _pending.Count >= Association.MaxAsyncOpsInvoked)
                {
                    return; 
}
When reverted back, Synchronous operations are restored(no errors during the send of the first image) if Asynchronous Operations are not negotiated in the Association.  Thanks again for looking at this.

**Comment by Corvin666** [comment]
I tried the fix @BobSter3000 suggested with servers built in DCM4CHE, Leadtools and our homegrown implementation. None of these three supported async ops and codepath was being hit. I observed no issues. 

I tried sending to fo-dicom based server and that one supported async ops, which was successfully executed. So, seems that silently returning kind of works.

One other thing I noticed - if I log everytime the codepath above is hit, _pending.Count never exceeds 1.

**Comment by aerik** [comment]
@BobSter3000  There are no consequences; it's just not intuitive.  If you follow through the code you'll observe what I explained in my Sept 21 comment.  The only potential zinger is what I mentioned there; "The original behavior works well as long as you receive a response, but sometimes you don't. My hacked up version of 1.x has a callback for ConnectionClosed to handle cases where no response was received bye DicomClient."

**Comment by Corvin666** [comment]
@aerik, are you saying that you can be in a situation when you don't receive a response, but connection is still open? 

Because if connection closes, ReadAndProcessPDU throws and you get notified..

**Comment by aerik** [comment]
@Corvin666 I want to be able to handle a closed connection programmatically, in a way that is cleaner than looking for ReadAndProcessPDU errors.

**Comment by anders9ustafsson** [comment]
I am going to try to summarize the concerns in this issue, please correct me if I am missing or misunderstanding something.

@BobSter3000 In your initial comment (Sep 20) you show the `FormatException` error from using 3.0.0-beta4. This error was separately reported (by @Corvin666) in issue #374, and it was fixed in PR #381.

According to [this](http://dicom.nema.org/medical/Dicom/current/output/chtml/part07/sect_D.3.3.3.html) section in the DICOM standard, async ops negotiation is optional. If not negotiated, async ops invoked and performed should be considered to be 1. If zero (0) async ops is explicitly negotiated, this should be interpreted as an unlimited amount of asynchronous operations.

Given this from the standard, I consider it reasonable that async ops invoked and performed are set to 1 if they are not explicitly negotiated, i.e. if `DicomClient.NegotiateAsyncOps` is **not** called. To negotiate unlimited asynchronousness, it suffices to call `DicomClient.NegotiateAsyncOps()` without arguments, since default values are 0 for both invoked and performed.

I realize that this is a breaking change in existing code, but I think it makes sense to require explicitness to adhere to the DICOM standard in this case.

So in conclusion, I would like to consider this issue as being resolved; PR #381 fixes the `FormatException` and multiple asynchronous operations should onwards be negotiated explicitly in order to comply with the DICOM standard.

If anyone has objections to my conclusion, please raise them soon enough. Otherwise I will close this issue shortly.

**Comment by Corvin666** [comment]
@anders9ustafsson, I think there is still an issue here with current codebase - if the peer reports supported async ops as 1 (after calling NegotiateAsyncOps() explicitly) and you are trying to send multiple files within a single association, the code fails. 

FormatException reporting by @BobSter3000 was masking the original exception. With fix from PR #381, it will execute

```
throw new DicomNetworkException(
                            "Cannot send messages since pending: {0} would exceed max async ops invoked: {1}",
                            _pending.Count,
                            Association.MaxAsyncOpsInvoked);
```

In my current app I made the fix suggested by @BobSter3000 in his Nov 20 comment and that allows running multiple C-STORE requests within same association either serially or asynchronously. 

Hope it makes sense.

**Comment by aerik** [comment]
@anders9ustafsson Hmm.. I haven't actually messed with the asynch ops - I had been assuming that everything sending to my PACS is sending synchronously.  My tests are only using synchronous C-STORE requests, and that's where I first noticed a problem.

The way the code used to be structured - returning silently - allowed for multiple synchronous sends.  I'm The code I'm using is a few months old, but without changing it I could not send multiple synchronous (serial) requests on the same association.  That's the part I'm concerned about.  I don't quite follow what you're saying about asynch operations, but I'm concerned that currently synchronous, serial operations (on the same association) are broken.

**Comment by Corvin666** [comment]
@aerik , I second that. With the "silent return" fix mentioned above, both sync (i.e. max async ops negotiated == 1) and async (max async ops > 1) operations are supported. Without the fix, sync ops are broken.

**Comment by anders9ustafsson** [comment]
Many thanks for your explanations, @Corvin666  and @aerik, I think I got the hang of it now. I will consider reverting to the "silent return" solution then. Just need to digest the idea a little :-)

**Comment by BobSter3000** [comment]
@anders9ustafsson Thanks for looking at this I really do appreciate it.  It looks like I am late with my response and @Corvin666 and @aerik has it covered.  Thanks again.

**Comment by ztegod** [comment]
@BobSter3000 Could you tell me the differences if i add the code ：
client.NegotiateAsyncOps(files.Length, files.Length);
In logs,it appears like all request to raise and all response to answer instead of one request with one response.Is it more faster to send images by adding the code by Async invoke and perform?

**Comment by BobSter3000** [comment]
@ztegod Yes, it generally is faster if Asynchronous operations is negotiated.
